# NeuroCamp Computer Science & Programming Day - Starter Notebook

1. Introduce the Jupyter Notebook, and run a "Hello World!" program in a notebook.
2. Variables and variable access.
3. I/O & load data from file.
4. Functions and modules. (We've used a function already!)
5. Loops and iteration.
6. Row field access. (Remember accessing values in a list by position or in a dictionary by key?)
7. Conditionals, and template strings for simple output formatting.
8. Math (change in measurement from A to B, percentage).
9. Write formatted output to screen.

## 1. Introduce the Jupyter Notebook, and run a "Hello World!" program in a notebook.

This file that you are working with is called a Jupyter Notebook. It allows you to add and run code in locations within the file called "cells". Cells that run code will have square brackets in front of them, and may also have a cell number within them.

For our work, we'll be using the Python language (version 3.7, currently). If you look at the top right of the window, you should see "Python 3". This is the language indicator, and shows you what language the file is expected to use.

At the top left of this pane, you'll see the toolbar. That lets you control what's happening in the file and within the current cell. Most importantly, the right-most dropdown allows you to change a cell from "Markdown" explanatory text mode to "Code" Python programming mode.

In the cell below, we use Python 3's built-in `print` function to write `Hello, World!` to the output console (displayed below the cell when it is run). We'll talk more about functions later. For now, just make sure you understand the mechanics of how the notebook interface works.

**Feel free to play with the code cell below, and see if you can change the word `"World"` to display your name.**

In [11]:
print("Hello, World!")

Hello, World!


## 2. Variables and variable access.

One thing we do frequently when writing programs is create variables to store data in memory that we can retrieve later. You can think of a "variable" as a box in which you put a piece of data called a "value".

You may have seen variables in algebra. When we do math, variables can only hold numbers.

When we write programs, the data that variables hold can have different types:

* text (called "strings")
* numbers (which can be "integer", "float", "double" and more)
* true/false (called "boolean")
* other types of single values
* structured and hierarchical data, such as lists, mappings (called "Dictionaries") and more (We'll see some of those later.)

Some languages require a different kind of box for each type of data you might want to put in. Python is gentler and allows us to put arbitrary types of data into the same variable. However, once you put a type of data into the variable, Python expects you to follow the rules of that type of data.

In the next cell, we're creating a couple string variables and then printing their values. Notice that the `print` function can accept variables representing the text ("string" value) to be written to the output.

**Once again, you should play with the code below and see if you can change the variable values.**

In [10]:
greeting = "Hello"
name = "World"

print(greeting + ", " + name + "!")

greeting = "Bienvenue"
name = "Arabella"

print(greeting + ", " + name + "!")

Hello, World!
Bienvenue, Arabella!


As you see, we can use variables in "expressions", like we did when we wrote `greeting + ", " + name + "!"`. That expression performs an operation called "concatenation" on the text given to it, providing us with a new string value that has all the text smashed together.

**_Aside_**: Note that we don't actually need the last `print` statement, since here in a notebook, entering only an expression (or even just a variable name by itself) as the last statement in a cell will evaluate that expression and print the value to the output area, as below. (Note, however, that the same is not true for values not at the end of the cell.)

In [8]:
greeting = "Hello"
name = "World"

greeting + ", " + name + "!"

greeting = "Bienvenue"
name = "Arabella"

greeting + ", " + name + "!"

'Bienvenue, Arabella!'

Expressions with number values also allow us to do mathematics, and more generally, expressions allow for manipulate the data stored in our variables.

In the below cell, we are creating two variables, then using them to create four more variables where the values are computed using expressions.

**Explore the code below to get a feel for how expressions work. For example, you might try to change the variable used in an expression, modify the mathematical operation used, or change the values of `a` and `b`.**

In [2]:
a = 1
b = 2
c = a + b
d = a - b
e = a * b
f = a / b

print("a =", a, ", b =", b, ", c =", c, ", d =", d, ", e =", e, ", f =", f)

a = 1 , b = 2 , c = 3 , d = -1 , e = 2 , f = 0.5


## 3. I/O & load data from file

When programming, "I/O" usually refers to "input/output", or the process of reading data from outside the program and delivering data outside the program.

We've already seen an I/O function: `print`. The `print` function is used to deliver string data from within the program to the screen, which lives outside the program.

Often, when working with I/O, we'll need to read from a file. In Python, files can be opened for reading using the `open` function, and lines can be read from the opened file using the `read` function.

Below, we read from our data file, located at `./data/data.csv` and print it to the screen.

In [4]:
opened_file = open('./data/data.csv')
contents = opened_file.read()
print(contents)
opened_file.close()

Subject,OSPAN_Group,OSPAN_Score,Distracter_easy,Distracter_hard
1,High,75,114.5,116.0
2,High,75,146.6,106.9
3,High,71,152.7,126.7
4,High,71,105.3,99.2
5,High,69,108.4,108.4
6,High,67,120.6,113.0
7,High,64,114.5,99.2
8,High,63,175.9,146.0
9,High,61,114.6,118.4
10,High,58,163.4,100.8
11,High,57,109.9,109.9
12,High,55,129.8,90.1
13,High,55,111.4,120.6
14,High,55,125.2,120.6
15,High,55,140.5,77.9
16,High,55,161.3,152.1
17,Low,52,120.6,123.7
18,Low,51,142.0,128.2
19,Low,51,117.6,129.8
20,Low,50,116.0,109.9
21,Low,48,151.1,152.7
22,Low,46,119.1,106.9
23,Low,41,142.0,135.9
24,Low,38,116.1,107.7
25,Low,38,135.9,109.9
26,Low,36,148.1,117.6
27,Low,22,163.4,154.2
28,Low,20,120.6,106.9
29,Low,19,106.1,107.7



This is great! We can make it a bit better, though.

Programmers often implement tools in computer languages to help them avoid problems. In this case, we have to remember to close the file after we're finished reading data from it. If we forget to do that, it can cause problems later on when trying to read the file, or even within the program.

To avoid this problem, the Python developers added a convenient way to write this same code that automatically closes the file when done. This uses something called "block syntax", which means that there is a collection of statements that are indented that represent all the things we want to do with the file while it is open. The block is started with a `with` statement.

In [5]:
with open('./data/data.csv') as opened_file:
    contents = opened_file.read()
    print(contents)

Subject,OSPAN_Group,OSPAN_Score,Distracter_easy,Distracter_hard
1,High,75,114.5,116.0
2,High,75,146.6,106.9
3,High,71,152.7,126.7
4,High,71,105.3,99.2
5,High,69,108.4,108.4
6,High,67,120.6,113.0
7,High,64,114.5,99.2
8,High,63,175.9,146.0
9,High,61,114.6,118.4
10,High,58,163.4,100.8
11,High,57,109.9,109.9
12,High,55,129.8,90.1
13,High,55,111.4,120.6
14,High,55,125.2,120.6
15,High,55,140.5,77.9
16,High,55,161.3,152.1
17,Low,52,120.6,123.7
18,Low,51,142.0,128.2
19,Low,51,117.6,129.8
20,Low,50,116.0,109.9
21,Low,48,151.1,152.7
22,Low,46,119.1,106.9
23,Low,41,142.0,135.9
24,Low,38,116.1,107.7
25,Low,38,135.9,109.9
26,Low,36,148.1,117.6
27,Low,22,163.4,154.2
28,Low,20,120.6,106.9
29,Low,19,106.1,107.7



It's great that we can see all the data, but right now, we're getting one variable `contents` that contains the entire file. Since this is tabular data, it would be nice to be able to get each row separately. For this, we need to look at a couple data structures: lists and Dictionaries, and to learn about...

## 4. Loops and iteration.

The list and dictionary data types were mentioned briefly above. Now we'll look at them more closely.

### Lists (also known as "Arrays")

A list in Python is an ordered collection of data. List values are written in Python as comma-separated lists enclosed in square brackets.

For example, in the next cell, we're defining a list containing the first 10 positive integers, and assigning that list-type value to the variable `my_list`.

In [7]:
my_list = [ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ]

my_list

[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

Lists can contain other types of data. For example, the following is a list of string values.

In [10]:
[ 'My', 'friend', 'ate', 'a', 'hot', 'dog', 'for', 'lunch' ]

['My', 'friend', 'ate', 'a', 'hot', 'dog', 'for', 'lunch']

If needed, we can put a variety of data types in the same list.

In [13]:
[ 'This', 'list', 'has', 0, 'and', 8, 'and', False ]

['This', 'list', 'has', 0, 'and', 8, 'and', False]

Values in a list are called "elements." To **access** the elements of a list, we can use the "index" of the element, which is a number representing the value's position within the list, starting from `0`.

For example, using `my_list` defined above, we can expect that the index `9`, our starting point, represents the first value in the list, `1`. Similarly, the index 5 refers to the fifth value in the list, which is `6`.

Let's see this in action.

**Play with the code below to explore lists. Here are a few things you might try:**

* What happens if you ask for the element at an index that is bigger than the length of the list?
* What happens if you ask for the element at an index that is negative?
* How would you retrieve the value 8 from the list?

In [22]:
my_list[4]

5

Were you able to answer the questions? OK! Let's apply this then.

**Now write a piece of code below that inspects the given array and reports the 3rd, 6th, and 9th elements by printing them to the screen.**

**Then, create a list of your top 3 favorite foods.**

In [32]:
words = [
    'This',
    'is',
    'my',
    'wonderful',
    'pet',
    'python',
    'Guido',
    'who',
    'enjoys',
    'life',
    'greatly'
]

In [34]:
# Enter your code here

# A. Print the 3rd, 6th and 9th elements of the given list of words to the output.

# B. Create a new list of your top 3 favorite foods, and print that to the output.


We can also modify list values after creation. To do this, we first select the element we want to change (as we did above), then use the equals sign to assign a new value to the list element.

In [36]:
words[3] = 'amazing'

words

['This',
 'is',
 'my',
 'amazing',
 'pet',
 'python',
 'Guido',
 'who',
 'enjoys',
 'life',
 'greatly']

There are many more ways to interact with lists. To learn more about lists and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/introduction.html#lists), and [the more detailed documentation here](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists).

## Tuples (like lists, but "immutable")

Variable types that allow the value to be changed after creation, like lists and strings and integers, are called "mutable" types.

There are also usually similar types that do not allow the value to be changed after creation. These are called "immutable" types, and in general, are referred to as "constants" rather than "variables". We'll look at one of those now.

An immutable type related to the python list is a tuple. Tuples are created the same way we create lists, but using parentheses instead of square brackets.

In [37]:
('my', 'tuple', 'value')

('my', 'tuple', 'value')

In [38]:
a_tuple = (2, 'values', 'here', False)

print(a_tuple)
print(a_tuple[2])

# Trying to change the value will generate an error
a_tuple[0] = 4

(2, 'values', 'here', False)
here


TypeError: 'tuple' object does not support item assignment

Tuples are sometimes used when creating dictionaries, which we'll look at next.

As with lists, there's much more we can do with tuples. To learn more about tuples and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/datastructures.html#tuples-and-sequences).

### Dictionaries (also known as "Hash Maps")

Dictionaries are data structures that associate a collection of keys to corresponding values.

For example, in the next cell, we are creating a dictionary that associates the key `'key_one'` with the value `'some value'` and the key `'key_two'` with the value `'some other value'`.

In [35]:
my_dict = {
    'key_one': 'some value',
    'key_two': 'some other value'
}

my_dict

{'key_one': 'some value', 'key_two': 'some other value'}

As with lists, dictionaries can contain different types of data, both for values and for keys, however all keys must use immutable variable types.

In [46]:
{
    3: True,
    'key': 2.71828,
    (1, 'this', False): {
        'another': 'dictionary',
        'as a': 'value',
        'with': 'a tuple',
        'as': 'a key',
        'but': 'not',
        'a list': 'as a key'
    },
    'lists': ['can', 'be', 'values', True]
}

{3: True,
 'key': 2.71828,
 (1, 'this', False): {'another': 'dictionary',
  'as a': 'value',
  'with': 'a tuple',
  'as': 'a key',
  'but': 'not',
  'a list': 'as a key'},
 'lists': ['can', 'be', 'values', True]}

In practice, however, dictionaries are most often used to look up values or functions, and because of that, keys are usually just string values.

To retrieve a value from a dictionary, we can use the key we associated with that value. For example, to retrieve the value `'some value'` from `my_dict` defined above, we would use `'key_one'`.

In [47]:
my_dict['key_one']

'some value'

We can also change the value associated with a key in a dictionary in the same way we do with lists.

For example, to change the value associated with `'key_two'` in `my_dict`, we first select that value, then use the equals sign to reassign a new value to the same key.

In [49]:
my_dict['key_two'] = 'some new value'

my_dict

{'key_one': 'some value', 'key_two': 'some new value'}

Since the keys in a dictionary are not just counting numbers, we may need to see what they are. We can do that by using the dictionary's `keys` function. (As an aside, functions on objects like a dictionary are usually referred to as "methods".)

In [50]:
my_dict.keys()

dict_keys(['key_one', 'key_two'])

Although less useful, we can also get the values of a dictionary using its `values` method, and its key-value pairs using the `items` method.

In [52]:
my_dict.values()

dict_values(['some value', 'some new value'])

In [54]:
my_dict.items()

dict_items([('key_one', 'some value'), ('key_two', 'some new value')])

As with lists and tuples, there is much, much more we can do with dictionaries. To learn more about dictionaries and how they work, you can review [the Python tutorial section for them here](https://docs.python.org/3/tutorial/datastructures.html#dictionaries).

**Play with the code below to explore dictionaries. Here are a few things you might try:**

* What happens if you ask for the element at using a key that is not in the dictionary?
* How would you retrieve the list of company divisions from the dictionary?
* How might you retrieve Bob Jones' telephone number from the dictionary?

In [72]:
company = {
    'divisions': [
        'Human Resources',
        'Research & Development',
        'Design',
        'Engineering',
        'Sales',
        'Marketing',
        'Finance',
        'Legal'
    ],
    'employees': [
        {
            'id': 125232,
            'name': 'Jenny Martin',
            'position': 'Sr. Software Engineer',
            'salary': 165000,
            'division': 'Engineering'
        },
        {
            'id': 653943,
            'name': 'Bob Jones',
            'position': 'Jr. Software Engineer',
            'salary': 59200,
            'division': 'Engineering'
        },
        {
            'id': 653943,
            'name': 'Althea Raines',
            'position': 'Chief Marketing Officer',
            'salary': 596400,
            'division': 'Sales'
        }
    ]
}

In [70]:
# Enter your code here


## 4. Functions and modules. (We've used a function already!)

Now that we know about these data structures, let's look at how to package our code into reusable chunks.

When we wrote the statement `print(contents)` above to see the contents of the file, we were calling a function. A function is a reusable, packaged set of code statements, potentially with some configuration parameters, that can be executed and may return a value. Executing a function with some data is called "calling" the function with "arguments".

Let's say we have a collection of statements that takes some numbers and provides a tuple of their sum and difference:

In [55]:
a = 3
b = 4

a_plus_b = a + b
a_minus_b = a - b
(a_plus_b, a_minus_b)

(7, -1)

To turn this into a function, we turn that code into a block like we did with the `with` statement above. This time, we'll use `def` to define a function. After `def`, we need to give the function a name, which we'll use to call it later. Following that, we need to specify the arguments (and optionally their types and default values). The statement ends in a colon (`:`), which tells Python to look for a block of code after that line.

In the block, we need to put the `return` keyword before the tuple to tell Python that we want to pass this value back once the function's execution is done.

Finally, we can call the new function using its name, and adding the values for `a` and `b` in parentheses.

In [57]:
def plus_and_minus( a:int, b:int ):
    a_plus_b = a + b
    a_minus_b = a - b
    return (a_plus_b, a_minus_b)

plus_and_minus(3, 4)

(7, -1)

**Try writing a function on your own. Create a function that takes 3 inputs (named whatever you like), and computes the difference of the first and second inputs divided by the third. Be sure to return that computed value. Then, try calling that function with a few sets of numbers to verify it does what you expect.**

In [73]:
# Enter your code here

# A. Define the function

# B. Call the function with several sets of arguments


Often, the amount of code in a project gets quite large, and some of that code ends up being used in a number of places. To make things more manageable, developers often put that code into functions and call those where needed. If that still gets complex to manage, a developer can create a "package", sometimes also called a "module".

A package is a collection of Python code that is created in its own project, and which we can import into our project when needed. The Python language developers often also use packages to provide language features that may not always be needed.

Conveniently, the Python language developers provided a package for working with CSV data files. You can review [the documentation for that package here](https://docs.python.org/3/library/csv.html).

To import that package, we use the `import` keyword.

In [58]:
import csv

We can now rewrite the code above to use the CSV package. The package provides a `reader` function (called with `csv.reader`). Let's see what that gives us.

In [60]:
with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    print(contents)

That doesn't help much. Thankfully, Python also provides a package for inspecting objects in a program. Let's see what that tells us.

In [66]:
import inspect

with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    print(contents)
    print('-----------')
    print(inspect.getdoc(contents))

-----------
CSV reader

Reader objects are responsible for reading and parsing tabular data
in CSV format.


This provides a bit more information, but not much. At this point, we [go to the documentation](https://docs.python.org/3/library/csv.html), and we see the following:

```
>>> import csv
>>> with open('eggs.csv', newline='') as csvfile:
...     spamreader = csv.reader(csvfile, delimiter=' ', quotechar='|')
...     for row in spamreader:
...         print(', '.join(row))
Spam, Spam, Spam, Spam, Spam, Baked Beans
Spam, Lovely Spam, Wonderful Spam
```

This shows us that we have to use a `for` statement, which we haven't seen yet. That brings us to...

## 5. Loops and iteration.

In computer science, the term "iteration" refers to making the computer repeat the same set of instructions multiple times. More commonly, a piece of code that implements an iteration is referred to as a "loop". While there are many kinds of loops, here, we'll look at the for loop.

In Python, a for loop begins with the keyword `for`, followed by an expression using the `in` keyword that defines the loop variable (`row` in the documentation example above), and the collection of data to select values from for the loop variable (`spamreader` in the documentation example above). The statement ends in a colon `:`, indicating to Python to expect a block following the statement. The block is the set of instructions to repeat for each value in the collection of data. In the documentation example above, the block contains a single statement: `print(', '.join(row))`.

In the code cell below, we've written a loop that iterates over the values in the tuple `(1, 4, 6, 5, 2)`, adds 3 to each, and prints the mathematical expression.

**Explore the code below to better understand how loops and iteration work. What happens when you...**

* Change the tuple to a list?
* Modify the statements in the block?

In [69]:
for num in (1, 4, 6, 5, 2):
    sum_value = num + 3
    print(num, "+ 3 =", sum_value)

1 + 3 = 4
4 + 3 = 7
6 + 3 = 9
5 + 3 = 8
2 + 3 = 5


There's much more we can learn about iteration, and Python has several unique features related to this concept, including generator expressions and list comprehensions, which are very powerful. To learn more about this, review [the Python tutorial section on for loops](https://docs.python.org/3/tutorial/controlflow.html#for-statements), [the more detailed Python documentation section on for statements](https://docs.python.org/3/reference/compound_stmts.html#for), and [the Python documentation on generators and generator expressions](https://docs.python.org/3/tutorial/classes.html#generators).

Let's rewrite our CSV file reading code to use the for loop.

In [74]:
with open('./data/data.csv') as opened_file:
    contents = csv.reader(opened_file)
    for row in contents:
        print(row)

['Subject', 'OSPAN_Group', 'OSPAN_Score', 'Distracter_easy', 'Distracter_hard']
['1', 'High', '75', '114.5', '116.0']
['2', 'High', '75', '146.6', '106.9']
['3', 'High', '71', '152.7', '126.7']
['4', 'High', '71', '105.3', '99.2']
['5', 'High', '69', '108.4', '108.4']
['6', 'High', '67', '120.6', '113.0']
['7', 'High', '64', '114.5', '99.2']
['8', 'High', '63', '175.9', '146.0']
['9', 'High', '61', '114.6', '118.4']
['10', 'High', '58', '163.4', '100.8']
['11', 'High', '57', '109.9', '109.9']
['12', 'High', '55', '129.8', '90.1']
['13', 'High', '55', '111.4', '120.6']
['14', 'High', '55', '125.2', '120.6']
['15', 'High', '55', '140.5', '77.9']
['16', 'High', '55', '161.3', '152.1']
['17', 'Low', '52', '120.6', '123.7']
['18', 'Low', '51', '142.0', '128.2']
['19', 'Low', '51', '117.6', '129.8']
['20', 'Low', '50', '116.0', '109.9']
['21', 'Low', '48', '151.1', '152.7']
['22', 'Low', '46', '119.1', '106.9']
['23', 'Low', '41', '142.0', '135.9']
['24', 'Low', '38', '116.1', '107.7']
['25'

This is much better! Now we see that the CSV package's `reader` function provides us with each row of the file as a list of string values.

## 6. Row field access. (Remember accessing values in a list by position or in a dictionary by key?)

Let's take each row apart, and print the values in the row with a vertical bar inbetween.

**Use what you've learned about lists to print each value in the `row` array separately, separated by the vertical bar character `|`.  For example, the row `['4', 'High', '71', '105.3', '99.2']` would print as:**

**`4 | High | 71 | 105.3 | 99.2`**

In [75]:
# Write code that prints each row's values separated by |.


This is nice, but referring to those values by their positions is a bit of a pain. It would be really nice to be able to refer to each row's column values by column name. This would require the CSV package to provide us a dictionary for each row. 

**Explore [the CSV package documentation](https://docs.python.org/3/library/csv.html) to see if you can find something that will do that.**

In [76]:
# Rewrite the code here using what you find in the CSV package documentation.


## 7. Conditionals, and template strings for simple output formatting.

...